vcr/log/log.go

```diff
@@ -0,0 +1,156 @@
+		},
+	}
+
+	cmd.Flags().StringVarP(&opts.InstanceID, "id", "i", "", "instance ID")
+	cmd.Flags().IntVarP(&opts.Limit, "tail", "l", 300, "prints the last N number of records")
+	cmd.Flags().StringVarP(&opts.ProjectName, "project-name", "p", "", "project name (must be used with instance-name flag)")
+	cmd.Flags().StringVarP(&opts.InstanceName, "instance-name", "n", "", "instance name (must be used with project-name flag)")
+	cmd.Flags().StringVarP(&opts.LogLevel, "log-level", "g", "", "filter for log level, e.g. info")
+	cmd.Flags().StringVarP(&opts.SourceType, "source-type", "s", "", "filter for source type e.g. application")
+
+	return cmd
+}
+
+func runLog(ctx context.Context, opts *Options) error {
+	io := opts.IOStreams()
+	if err := cmdutil.ValidateFlags(opts.InstanceID, opts.InstanceName, opts.ProjectName); err != nil {
+		return fmt.Errorf("failed to validate flags: %w", err)
+	}
+
+	inst, err := getInstance(ctx, opts)
+	if err != nil {
+		return fmt.Errorf("failed to get instance: %w", err)
+	}
+
+	opts.InstanceID = inst.ID
+
+	ticker := time.NewTicker(TickerInterval)
+	defer ticker.Stop()
+	lastTimestamp := time.Time{}
+
+	interrupt := make(chan os.Signal, 1)
+	signal.Notify(interrupt, os.Interrupt, syscall.SIGTERM)
+
+Loop:
+	for {
+		select {
+		case <-ticker.C:
+			fetchLogs(io, opts, &lastTimestamp)
+		case <-interrupt:
+			fmt.Println("Interrupt received, stopping...")
+			break Loop
+		}
+	}
+
+	return nil
+}
+
+func fetchLogs(out *iostreams.IOStreams, opts *Options, lastTimestamp *time.Time) {
+	c := out.ColorScheme()
+	ctx, cancel := context.WithDeadline(context.Background(), time.Now().Add(opts.Timeout()))
+	defer cancel()
+	logs, err := opts.Datastore().ListLogsByInstanceID(ctx, opts.InstanceID, opts.Limit, *lastTimestamp)
+	if err != nil {
+		fmt.Fprintf(out.ErrOut, "%s Error fetching logs: %v\n", c.WarningIcon(), err)
+		return
+	}
+
+	for i := len(logs) - 1; i >= 0; i-- {
+		log := logs[i]
+		printLogs(out, opts, log)
+		*lastTimestamp = log.Timestamp
+	}
+}
+
+func getInstance(ctx context.Context, opts *Options) (api.Instance, error) {
+	if opts.InstanceID != "" {
+		inst, err := opts.Datastore().GetInstanceByID(ctx, opts.InstanceID)
+		if err != nil {
+			if errors.Is(err, api.ErrNotFound) {
+				return api.Instance{}, fmt.Errorf("instance %q not found", opts.InstanceID)
+			}
+			return api.Instance{}, err
+		}
+		return inst, nil
+	}
+	inst, err := opts.Datastore().GetInstanceByProjectAndInstanceName(ctx, opts.ProjectName, opts.InstanceName)
+	if err != nil {
+		if errors.Is(err, api.ErrNotFound) {
+			return api.Instance{}, fmt.Errorf("instance with project name %q and instance name %q not found", opts.ProjectName, opts.InstanceName)
+		}
+		return api.Instance{}, err
+	}
+	return inst, nil
+}
+
+func printLogs(out *iostreams.IOStreams, opts *Options, log api.Log) {
+	switch {
+	case opts.SourceType != "" && opts.LogLevel != "":
+		if opts.SourceType != log.SourceType || opts.LogLevel != log.LogLevel {
+			return
```

Comment:
remember that log level shows all levels that a more severe as well. ie. if we set log level to info that means we do warn, error, fatal as well, but we do not show debug and trace